File: artifacts/api-server/src/routes/sessions.ts
// Rutas de sesiones de competencia: el docente crea una sesion con un codigo unico
// Los estudiantes se unen con el link compartible y compiten en el mismo reto
import { Router } from "express";
import { db } from "@workspace/db";
import {
  sesionesCompetenciaTable,
  participantesSesionTable,
  retosTable,
  perfilesTable,
} from "@workspace/db";
import { eq, and, desc } from "drizzle-orm";
import { requireAuth, type AuthRequest } from "../lib/auth";

const router = Router();

function generateCode(): string {
  const chars = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
  let code = "";
  for (let i = 0; i < 6; i++) {
    code += chars[Math.floor(Math.random() * chars.length)];
  }
  return code;
}

// POST /api/sessions — docente crea una sesion de competencia
router.post("/sessions", requireAuth, async (req: AuthRequest, res) => {
  try {
    if (req.user!.rol !== "docente") {
      res.status(403).json({ error: "forbidden", message: "Solo los docentes pueden crear sesiones" });
      return;
    }
    const { nombre, id_reto } = req.body;
    if (!nombre?.trim() || !id_reto) {
      res.status(400).json({ error: "validation_error", message: "nombre e id_reto son requeridos" });
      return;
    }

    let codigo = generateCode();
    for (let i = 0; i < 10; i++) {
      const [existing] = await db.select().from(sesionesCompetenciaTable)
        .where(eq(sesionesCompetenciaTable.codigo, codigo));
      if (!existing) break;
      codigo = generateCode();
    }

    const [sesion] = await db.insert(sesionesCompetenciaTable).values({
      codigo,
      nombre: nombre.trim(),
      id_reto: Number(id_reto),
      id_docente: req.user!.id,
      activa: true,
    }).returning();

    res.status(201).json({ ...sesion, link: `/competir/${codigo}` });
  } catch (err) {
    req.log.error({ err }, "Error al crear sesion");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /api/sessions — docente obtiene todas sus sesiones
router.get("/sessions", requireAuth, async (req: AuthRequest, res) => {
  try {
    if (req.user!.rol !== "docente") {
      res.status(403).json({ error: "forbidden", message: "Solo los docentes pueden ver sus sesiones" });
      return;
    }

    const sesiones = await db.select().from(sesionesCompetenciaTable)
      .where(eq(sesionesCompetenciaTable.id_docente, req.user!.id))
      .orderBy(desc(sesionesCompetenciaTable.creado_en));

    const sesionesConInfo = await Promise.all(sesiones.map(async (s) => {
      const participantes = await db.select().from(participantesSesionTable)
        .where(eq(participantesSesionTable.id_sesion, s.id));
      const [reto] = await db.select({ nombre: retosTable.nombre, tipo_juego: retosTable.tipo_juego })
        .from(retosTable).where(eq(retosTable.id, s.id_reto));
      return {
        ...s,
        participantes_count: participantes.length,
        reto_nombre: reto?.nombre || "—",
        reto_tipo: reto?.tipo_juego || "",
      };
    }));

    res.json(sesionesConInfo);
  } catch (err) {
    req.log.error({ err }, "Error al obtener sesiones");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /api/sessions/:code — info de sesion publica (para que el estudiante se una)
router.get("/sessions/:code", requireAuth, async (req: AuthRequest, res) => {
  try {
    const [sesion] = await db.select().from(sesionesCompetenciaTable)
      .where(eq(sesionesCompetenciaTable.codigo, (req.params.code as string).toUpperCase()));

    if (!sesion) {
      res.status(404).json({ error: "not_found", message: "Sesion no encontrada" });
      return;
    }

    const [reto] = await db.select().from(retosTable).where(eq(retosTable.id, sesion.id_reto));
    const [docente] = await db.select({ nombre: perfilesTable.nombre })
      .from(perfilesTable).where(eq(perfilesTable.id, sesion.id_docente));
    const participantes = await db.select().from(participantesSesionTable)
      .where(eq(participantesSesionTable.id_sesion, sesion.id));

    res.json({
      ...sesion,
      reto,
      docente_nombre: docente?.nombre || "Docente",
      participantes_count: participantes.length,
    });
  } catch (err) {
    req.log.error({ err }, "Error al obtener sesion");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// POST /api/sessions/:code/result — estudiante registra su resultado en la sesion
router.post("/sessions/:code/result", requireAuth, async (req: AuthRequest, res) => {
  try {
    const code = (req.params.code as string).toUpperCase();
    const [sesion] = await db.select().from(sesionesCompetenciaTable)
      .where(eq(sesionesCompetenciaTable.codigo, code));

    if (!sesion) {
      res.status(404).json({ error: "not_found", message: "Sesion no encontrada" });
      return;
    }
    if (!sesion.activa) {
      res.status(400).json({ error: "session_closed", message: "La sesion ya no esta activa" });
      return;
    }

    const { puntuacion, tiempo_total, respuestas_correctas, total_preguntas } = req.body;
    const [existing] = await db.select().from(participantesSesionTable)
      .where(and(
        eq(participantesSesionTable.id_sesion, sesion.id),
        eq(participantesSesionTable.id_usuario, req.user!.id),
      ));

    if (existing) {
      const [updated] = await db.update(participantesSesionTable).set({
        puntuacion: Number(puntuacion || 0),
        tiempo_total: Number(tiempo_total || 0),
        respuestas_correctas: Number(respuestas_correctas || 0),
        total_preguntas: Number(total_preguntas || 0),
        completado: true,
      }).where(eq(participantesSesionTable.id, existing.id)).returning();
      res.json(updated);
    } else {
      const [nuevo] = await db.insert(participantesSesionTable).values({
        id_sesion: sesion.id,
        id_usuario: req.user!.id,
        puntuacion: Number(puntuacion || 0),
        tiempo_total: Number(tiempo_total || 0),
        respuestas_correctas: Number(respuestas_correctas || 0),
        total_preguntas: Number(total_preguntas || 0),
        completado: true,
      }).returning();
      res.status(201).json(nuevo);
    }
  } catch (err) {
    req.log.error({ err }, "Error al registrar resultado de sesion");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /api/sessions/:code/ranking — ranking de la sesion de competencia
router.get("/sessions/:code/ranking", requireAuth, async (req: AuthRequest, res) => {
  try {
    const code = (req.params.code as string).toUpperCase();
    const [sesion] = await db.select().from(sesionesCompetenciaTable)
      .where(eq(sesionesCompetenciaTable.codigo, code));

    if (!sesion) {
      res.status(404).json({ error: "not_found", message: "Sesion no encontrada" });
      return;
    }

    const participantes = await db.select().from(participantesSesionTable)
      .where(eq(participantesSesionTable.id_sesion, sesion.id))
      .orderBy(desc(participantesSesionTable.puntuacion));

    const rankingConPerfiles = await Promise.all(participantes.map(async (p, index) => {
      const [perfil] = await db.select({
        nombre: perfilesTable.nombre,
        usuario: perfilesTable.usuario,
        grado_bachillerato: perfilesTable.grado_bachillerato,
      }).from(perfilesTable).where(eq(perfilesTable.id, p.id_usuario));

      return {
        posicion: index + 1,
        usuario_id: p.id_usuario,
        nombre: perfil?.nombre || "Estudiante",
        usuario: perfil?.usuario || "",
        grado_bachillerato: perfil?.grado_bachillerato ?? null,
        puntuacion: p.puntuacion,
        tiempo_total: p.tiempo_total,
        respuestas_correctas: p.respuestas_correctas,
        total_preguntas: p.total_preguntas,
        completado: p.completado,
      };
    }));

    res.json(rankingConPerfiles);
  } catch (err) {
    req.log.error({ err }, "Error al obtener ranking de sesion");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /api/sessions/:code/report — informe detallado para el docente
router.get("/sessions/:code/report", requireAuth, async (req: AuthRequest, res) => {
  try {
    if (req.user!.rol !== "docente") {
      res.status(403).json({ error: "forbidden", message: "Solo los docentes pueden ver el informe" });
      return;
    }

    const code = (req.params.code as string).toUpperCase();
    const [sesion] = await db.select().from(sesionesCompetenciaTable)
      .where(and(
        eq(sesionesCompetenciaTable.codigo, code),
        eq(sesionesCompetenciaTable.id_docente, req.user!.id),
      ));

    if (!sesion) {
      res.status(404).json({ error: "not_found", message: "Sesion no encontrada o no eres el creador" });
      return;
    }

    const [reto] = await db.select().from(retosTable).where(eq(retosTable.id, sesion.id_reto));
    const participantes = await db.select().from(participantesSesionTable)
      .where(eq(participantesSesionTable.id_sesion, sesion.id))
      .orderBy(desc(participantesSesionTable.puntuacion));

    const participantesConPerfiles = await Promise.all(participantes.map(async (p) => {
      const [perfil] = await db.select({
        nombre: perfilesTable.nombre,
        usuario: perfilesTable.usuario,
        grado_bachillerato: perfilesTable.grado_bachillerato,
      }).from(perfilesTable).where(eq(perfilesTable.id, p.id_usuario));
      return { ...p, perfil };
    }));

    const total = participantes.length;
    const completaron = participantes.filter((p) => p.completado).length;
    const promPuntuacion = total > 0
      ? Math.round(participantes.reduce((s, p) => s + p.puntuacion, 0) / total) : 0;
    const promPrecision = total > 0
      ? Math.round(participantes.reduce((s, p) =>
          s + (p.total_preguntas > 0 ? (p.respuestas_correctas / p.total_preguntas) * 100 : 0), 0) / total)
      : 0;

    res.json({
      sesion: { ...sesion, reto },
      estadisticas: { totalParticipantes: total, completaron, promPuntuacion, promPrecision },
      participantes: participantesConPerfiles,
    });
  } catch (err) {
    req.log.error({ err }, "Error al obtener informe de sesion");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// PATCH /api/sessions/:code/toggle — docente activa o desactiva la sesion
router.patch("/sessions/:code/toggle", requireAuth, async (req: AuthRequest, res) => {
  try {
    if (req.user!.rol !== "docente") {
      res.status(403).json({ error: "forbidden" });
      return;
    }
    const code = (req.params.code as string).toUpperCase();
    const [sesion] = await db.select().from(sesionesCompetenciaTable)
      .where(and(
        eq(sesionesCompetenciaTable.codigo, code),
        eq(sesionesCompetenciaTable.id_docente, req.user!.id),
      ));
    if (!sesion) {
      res.status(404).json({ error: "not_found" });
      return;
    }
    const [updated] = await db.update(sesionesCompetenciaTable)
      .set({ activa: !sesion.activa })
      .where(eq(sesionesCompetenciaTable.id, sesion.id))
      .returning();
    res.json(updated);
  } catch (err) {
    req.log.error({ err }, "Error al alternar sesion");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

export default router;

File: artifacts/api-server-python/routers/sessions.py
import random
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from db import get_db, fetchone, fetchall, execute
from auth_utils import get_current_user
from datetime import datetime

router = APIRouter()

CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"


def generate_code() -> str:
    return "".join(random.choices(CHARS, k=6))


def dt(v):
    if isinstance(v, datetime):
        return v.isoformat()
    return v


def serialize_row(row):
    if row is None:
        return None
    r = dict(row)
    for k, v in r.items():
        if isinstance(v, datetime):
            r[k] = v.isoformat()
    return r


class CreateSessionBody(BaseModel):
    nombre: str
    id_reto: int


@router.post("/sessions", status_code=201)
def create_session(body: CreateSessionBody, user=Depends(get_current_user)):
    if user["rol"] != "docente":
        raise HTTPException(403, "Solo los docentes pueden crear sesiones")
    if not body.nombre or not body.nombre.strip() or not body.id_reto:
        raise HTTPException(400, "nombre e id_reto son requeridos")

    with get_db() as conn:
        codigo = generate_code()
        for _ in range(10):
            existing = fetchone(conn, "SELECT id FROM sesiones_competencia WHERE codigo = %s", (codigo,))
            if not existing:
                break
            codigo = generate_code()

        with conn.cursor() as cur:
            cur.execute(
                "INSERT INTO sesiones_competencia (codigo, nombre, id_reto, id_docente, activa) VALUES (%s, %s, %s, %s, 1)",
                (codigo, body.nombre.strip(), body.id_reto, user["id"]),
            )
            conn.commit()
            sesion_id = cur.lastrowid

        sesion = fetchone(conn, "SELECT * FROM sesiones_competencia WHERE id = %s", (sesion_id,))

    r = serialize_row(sesion)
    r["link"] = f"/competir/{codigo}"
    return r


@router.get("/sessions")
def get_sessions(user=Depends(get_current_user)):
    if user["rol"] != "docente":
        raise HTTPException(403, "Solo los docentes pueden ver sus sesiones")

    with get_db() as conn:
        sesiones = fetchall(conn,
            "SELECT * FROM sesiones_competencia WHERE id_docente = %s ORDER BY creado_en DESC",
            (user["id"],))
        result = []
        for s in sesiones:
            s = dict(s)
            participantes = fetchall(conn, "SELECT id FROM participantes_sesion WHERE id_sesion = %s", (s["id"],))
            reto = fetchone(conn, "SELECT nombre, tipo_juego FROM retos WHERE id = %s", (s["id_reto"],))
            r = serialize_row(s)
            r["participantes_count"] = len(participantes)
            r["reto_nombre"] = reto["nombre"] if reto else "—"
            r["reto_tipo"] = reto["tipo_juego"] if reto else ""
            result.append(r)
    return result


@router.get("/sessions/{code}")
def get_session(code: str, user=Depends(get_current_user)):
    with get_db() as conn:
        sesion = fetchone(conn,
            "SELECT * FROM sesiones_competencia WHERE codigo = %s",
            (code.upper(),))
        if not sesion:
            raise HTTPException(404, "Sesion no encontrada")
        reto = fetchone(conn, "SELECT * FROM retos WHERE id = %s", (sesion["id_reto"],))
        docente = fetchone(conn, "SELECT nombre FROM perfiles WHERE id = %s", (sesion["id_docente"],))
        participantes = fetchall(conn, "SELECT id FROM participantes_sesion WHERE id_sesion = %s", (sesion["id"],))

    r = serialize_row(sesion)
    r["reto"] = serialize_row(reto)
    r["docente_nombre"] = docente["nombre"] if docente else "Docente"
    r["participantes_count"] = len(participantes)
    return r


class SessionResultBody(BaseModel):
    puntuacion: Optional[float] = 0
    tiempo_total: Optional[int] = 0
    respuestas_correctas: Optional[int] = 0
    total_preguntas: Optional[int] = 0


@router.post("/sessions/{code}/result")
def session_result(code: str, body: SessionResultBody, user=Depends(get_current_user)):
    with get_db() as conn:
        sesion = fetchone(conn,
            "SELECT * FROM sesiones_competencia WHERE codigo = %s",
            (code.upper(),))
        if not sesion:
            raise HTTPException(404, "Sesion no encontrada")
        if not sesion["activa"]:
            raise HTTPException(400, "La sesion ya no esta activa")

        existing = fetchone(conn,
            "SELECT * FROM participantes_sesion WHERE id_sesion = %s AND id_usuario = %s",
            (sesion["id"], user["id"]))

        if existing:
            with conn.cursor() as cur:
                cur.execute(
                    "UPDATE participantes_sesion SET puntuacion=%s, tiempo_total=%s, respuestas_correctas=%s, total_preguntas=%s, completado=1 WHERE id=%s",
                    (body.puntuacion, body.tiempo_total, body.respuestas_correctas, body.total_preguntas, existing["id"]),
                )
                conn.commit()
            updated = fetchone(conn, "SELECT * FROM participantes_sesion WHERE id = %s", (existing["id"],))
            return serialize_row(updated)
        else:
            with conn.cursor() as cur:
                cur.execute(
                    "INSERT INTO participantes_sesion (id_sesion, id_usuario, puntuacion, tiempo_total, respuestas_correctas, total_preguntas, completado) VALUES (%s,%s,%s,%s,%s,%s,1)",
                    (sesion["id"], user["id"], body.puntuacion, body.tiempo_total, body.respuestas_correctas, body.total_preguntas),
                )
                conn.commit()
                nuevo_id = cur.lastrowid
            nuevo = fetchone(conn, "SELECT * FROM participantes_sesion WHERE id = %s", (nuevo_id,))
            return serialize_row(nuevo)


@router.get("/sessions/{code}/ranking")
def session_ranking(code: str, user=Depends(get_current_user)):
    with get_db() as conn:
        sesion = fetchone(conn, "SELECT * FROM sesiones_competencia WHERE codigo = %s", (code.upper(),))
        if not sesion:
            raise HTTPException(404, "Sesion no encontrada")

        participantes = fetchall(conn,
            "SELECT * FROM participantes_sesion WHERE id_sesion = %s ORDER BY puntuacion DESC",
            (sesion["id"],))
        result = []
        for i, p in enumerate(participantes):
            perfil = fetchone(conn,
                "SELECT nombre, usuario, grado_bachillerato FROM perfiles WHERE id = %s",
                (p["id_usuario"],))
            result.append({
                "posicion": i + 1,
                "usuario_id": p["id_usuario"],
                "nombre": perfil["nombre"] if perfil else "Estudiante",
                "usuario": perfil["usuario"] if perfil else "",
                "grado_bachillerato": perfil["grado_bachillerato"] if perfil else None,
                "puntuacion": p["puntuacion"],
                "tiempo_total": p.get("tiempo_total"),
                "respuestas_correctas": p.get("respuestas_correctas"),
                "total_preguntas": p.get("total_preguntas"),
                "completado": bool(p.get("completado")),
            })
    return result


@router.get("/sessions/{code}/report")
def session_report(code: str, user=Depends(get_current_user)):
    if user["rol"] != "docente":
        raise HTTPException(403, "Solo los docentes pueden ver el informe")

    with get_db() as conn:
        sesion = fetchone(conn,
            "SELECT * FROM sesiones_competencia WHERE codigo = %s AND id_docente = %s",
            (code.upper(), user["id"]))
        if not sesion:
            raise HTTPException(404, "Sesion no encontrada o no eres el creador")

        reto = fetchone(conn, "SELECT * FROM retos WHERE id = %s", (sesion["id_reto"],))
        participantes = fetchall(conn,
            "SELECT * FROM participantes_sesion WHERE id_sesion = %s ORDER BY puntuacion DESC",
            (sesion["id"],))
        participantes_con_perfil = []
        for p in participantes:
            perfil = fetchone(conn,
                "SELECT nombre, usuario, grado_bachillerato FROM perfiles WHERE id = %s",
                (p["id_usuario"],))
            row = serialize_row(p)
            row["perfil"] = serialize_row(perfil)
            participantes_con_perfil.append(row)

    total = len(participantes)
    completaron = sum(1 for p in participantes if p.get("completado"))
    prom_puntuacion = round(sum(p["puntuacion"] for p in participantes) / total) if total else 0
    prom_precision = round(
        sum((p["respuestas_correctas"] / p["total_preguntas"] * 100) if p.get("total_preguntas") else 0
            for p in participantes) / total
    ) if total else 0

    return {
        "sesion": {**serialize_row(sesion), "reto": serialize_row(reto)},
        "estadisticas": {
            "totalParticipantes": total,
            "completaron": completaron,
            "promPuntuacion": prom_puntuacion,
            "promPrecision": prom_precision,
        },
        "participantes": participantes_con_perfil,
    }


@router.patch("/sessions/{code}/toggle")
def toggle_session(code: str, user=Depends(get_current_user)):
    if user["rol"] != "docente":
        raise HTTPException(403, "forbidden")

    with get_db() as conn:
        sesion = fetchone(conn,
            "SELECT * FROM sesiones_competencia WHERE codigo = %s AND id_docente = %s",
            (code.upper(), user["id"]))
        if not sesion:
            raise HTTPException(404, "not_found")

        nuevo = not bool(sesion["activa"])
        execute(conn, "UPDATE sesiones_competencia SET activa = %s WHERE id = %s",
                (1 if nuevo else 0, sesion["id"]))
        updated = fetchone(conn, "SELECT * FROM sesiones_competencia WHERE id = %s", (sesion["id"],))
    return serialize_row(updated)
